README.md
# What is Neon Machine

## Introduction

Neon Machine is a Virtual Machine \(VM\) built on the Love2D Lua Engine. It is similar to the ComputerCraft mod for Minecraft and essentially consists of a console/terminal.

If you'd like to try it out, or maybe even help developing it, head over to [Getting started](getting-started/) and follow the instructions.

## Contributors/credits

People who have helped to make this project, or otherwise deserve a shout-out

{% page-ref page="contributors-credits.md" %}



changelog.md
---
description: All changes since 29/10/2018
---

# Changelog

## 0.1.0 Alpha - 29/10/2018

### Added

* Core functionality of the program \(won't go into details, shouldn't matter at this point\)
* Documentation using GitBook



kernel/rom/core/gui.lua
local setmetatable, table, assert, pairs = setmetatable, table, assert, pairs

require("kernel.rom.core.libs.tree")
local Tree = tree.Tree

---@module gui
module("gui")

---@class GUI
---@field List table
---@field width number
---@field height number
---@field hitboxes Tree
GUI = {
    List = {}, -- list of instances
}
GUI_mt = {}

---new
---@param _width number
---@param _height number
---@return GUI
function GUI.new(_width, _height)
    local o = setmetatable({
        width = _width or 200,
        height = _height or 150,
        hitboxes = Tree.new(2),
    }, GUI_mt)
    table.insert(GUI.List, o)
    return o
end

---addHitbox
---@param _id string
---@param _onClick function
---@param _x number
---@param _y number
---@param _width number
---@param _height number
---@param _zIndex number
---@return Hitbox
function GUI:addHitbox(_id, _onClick, _onHover, _x, _y, _width, _height, _zIndex)
    assert(_id, "Hitboxes need an id")

    ---@class Hitbox
    ---@field id string
    ---@field onClick function
    ---@field onHover function
    ---@field x number
    ---@field y number
    ---@field z number
    ---@field width number
    ---@field height number
    local o = { -- Prototype/template
        id = _id,
        onClick = _onClick or function() end,
        onHover = _onHover or function() end,
        x = _x or 0,
        y = _y or 0,
        z = _zIndex or 1,
        width = _width or 1,
        height = _height or 1,
    }

    self.hitboxes:insert(o)
    return o
end

---removeHitbox
---@param _id string
---@return Hitbox
function GUI:removeHitbox(_id)
    return self.hitboxes:removeLeaf(_id)
end

---getHitboxAtPoint
---@param x number
---@param y number
---@return Hitbox
function GUI:getHitboxAtPoint(x, y)
    ---@type table
    local res = self.hitboxes:find(x, y)
    ---@type Hitbox
    local b
    ---@param v Hitbox
    for k,v in pairs(res) do
        if v.z and (not b or v.z > b.z) then
            b = v
        end
    end
    return b
end

GUI_mt.__index = GUI


kernel/rom/core/libs/tree.lua
local setmetatable, table, pairs, error = setmetatable, table, pairs, error

---@module tree
module("tree")

-- Classes (and their metatables)
---@class Tree
---@field maxLeaves number
---@field leaves table
---@field root Space
Tree = {}
Tree_mt = {}
---@class Space
Space = {}
Space_mt = {}
---@class Leaf
Leaf = {}
Leaf_mt = {}


-- Local functions
---getPropertiesFromLevel
---@param o Space | Leaf
---@param level number
---@return number, number
local function getPropertiesFromLevel(o, level) -- Only usable by spaces and leaves
    if (level%2 == 1) then
        return o.x, o.width
    else
        return o.y, o.height
    end
end

---intersects
---@param x1 number
---@param l1 number
---@param x2 number
---@param l2 number
---@return boolean
function intersects(x1, l1, x2, l2) -- A one-dimensional intersection checker
    if l1 == 0 then l1 = nil end
    return (x1 >= x2 and x1 <= x2+l2) or (l1 and x1+l1 <= x2 and x1+l1 <= x2+l2)
end


-- Class constructors and methods
---new
---@param _maxLeaves number
---@param _leaves number
---@return Tree
function Tree.new(_maxLeaves, _leaves)
    local o = {
        maxLeaves = _maxLeaves or 3,
        leaves = _leaves or {},
    }
    setmetatable(o, Tree_mt)
    return o
end

---insert
---@param leaf Leaf
---@param parent Space
function Tree:insert(leaf, parent)
    if not parent then -- First insertion (non-recursive)
        leaf = Leaf.new(leaf)
        if leaf.id then
            self.leaves[leaf.id] = leaf
        else
            table.insert(self.leaves, leaf)
        end
        if not self.root then -- Is the tree empty?
            self.root = Space.new({
                x = leaf.x,
                y = leaf.y,
                width = leaf.width,
                height = leaf.height,
            })
            table.insert(self.root.leaves, leaf)
            return
        end
        parent = self.root
    end

    -- Recursive stuff

    -- set space size
    if leaf.x < parent.x then
        parent.width = parent.width + (parent.x-leaf.x)
        parent.x = leaf.x
    end
    if leaf.y < parent.y then
        parent.height = parent.height + (parent.y-leaf.y)
        parent.y = leaf.y
    end
    if leaf.x+leaf.width > parent.x+parent.width then parent.width = leaf.width - (parent.x - leaf.x) end
    if leaf.y+leaf.height > parent.y+parent.height then parent.height = leaf.height - (parent.y - leaf.y) end

    if not parent.leaves then -- insert into child spaces
        local start, length = getPropertiesFromLevel(leaf, parent.level)
        local side = parent:compare(start, length) and "right" or "left"

        if not parent[side] then -- make side if it doesn't exist
            parent[side] = Space.new({
                x = leaf.x,
                y = leaf.y,
                width = leaf.width,
                height = leaf.height,
                level = parent.level+1
            })
        end
        -- add to side
        self:insert(leaf, parent[side])
    else -- insert into leaves
        table.insert(parent.leaves, leaf)

        if #parent.leaves > self.maxLeaves then
            local leaves = parent.leaves
            parent.leaves = nil
            for k,v in pairs(leaves) do
                self:insert(v, parent)
            end
        end
    end
end

---removeLeaf
---@param _id string
---@return Leaf the removed leaf
function Tree:removeLeaf(_id)
    local leaf = self.leaves[_id]
    if leaf then
        self.root:removeLeaf(leaf)
    end
    return leaf
end

function Tree:find(x, y)
    if not self.root then
        return {}
    end
    return self.root:find(x, y)
end

function Tree:findArea(x, width, y, height)

end

Tree_mt.__index = Tree


---new
---@param o table
---@return Space
function Space.new(o)
    local n = {
        x = 0,
        y = 0,
        width = 1,
        height = 1,
        level = 1,
        leaves = {},
    }
    for k,v in pairs(o) do
        n[k] = v
    end
    return setmetatable(n, Space_mt)
end

---compare
---@param p1 number
---@param p2 number
---@return boolean true if the middle of the given points is greater than or equal to the middle of the space
function Space:compare(p1, p2) -- Compare middles
    local start, length = getPropertiesFromLevel(self, self.level)
    return p1 + p2/2 >= start + length/2
end

---intersects
---@param x number
---@param width number
---@param y number
---@param height number
---@return boolean true if the given square (or point) intersects with the space
function Space:intersects(x, width, y, height) -- note that all params are optional
    return (not x or intersects(x, width, self.x, self.width)) and (not y or intersects(y, height, self.y, self.height))
end

function Space:contains(x, y)
    return intersects(x, 0, self.x, self.width) and intersects(y, 0, self.y, self.height)
end

function Space:find(x, y)
    local res = {}
    if self.leaves then
        ---@param v Leaf
        for k,v in pairs(self.leaves) do
            if v:contains(x, y) then
                table.insert(res, v)
            end
        end
    else
        if self.left and self.left:contains(x, y) then
            for k,v in pairs(self.left:find(x,y)) do
                table.insert(res, v)
            end
        end
        if self.right and self.right:contains(x, y) then
            for k,v in pairs(self.right:find(x,y)) do
                table.insert(res, v)
            end
        end
    end

    return res
end

---removeLeaf
---@param leaf Leaf
function Space:removeLeaf(leaf)
    if self.leaves then
        for k,v in pairs(self.leaves) do
            if v == leaf then
                table.remove(self.leaves, k)
            end
        end
    else
        if self.left and self.left:contains(leaf.x, leaf.y) then
            self.left.removeLeaf()
        end
        if self.right and self.right:contains(leaf.x, leaf.y) then
            self.right.removeLeaf()
        end
    end
end

Space_mt.__index = Space


---new
---@param o table
---@return Leaf
function Leaf.new(o)
    ---@type Leaf
    local n = {
        x = 0,
        y = 0,
        width = 1,
        height = 1,
    }
    for k,v in pairs(o) do
        n[k] = v
    end
    return setmetatable(n, Leaf_mt)
end

---contains
---@param x number
---@param y number
---@return Leaf
function Leaf:contains(x, y)
    return intersects(x, 0, self.x, self.width) and intersects(y, 0, self.y, self.height)
end

Leaf_mt.__index = Leaf


main.lua
local CREDITS = {
	"LeDark Lua a.k.a. Laurynas Šuopys",
	"Andreas \"Andy\" Clausen",
}

local socket = require("socket")
local floor=math.floor
local ceil=math.ceil
local twofivefive = 1/255
function round( num )
	if num >= 0 then return floor( num+.5 ) 
	else return ceil( num-.5 ) end
end
local canvas
local function map( n, start1, stop1, start2, stop2 )
	return (n - start1) / (stop1 - start1) * (stop2 - start2) + start2
end
math.map = map
local win = {
	[ "width" ] = 200;
	[ "height" ] = 150;
	[ "scale" ] = 4;
}
local blueCutoff = 0
local blueAmount = 5
local MAKE_TRUE_SIZE = 8
local MAX_RAM = 16384
local USED_RAM = 0
local MAX_STORAGE = 131072
local USED_STORAGE = 0
local CPU_SPEED = 1

local LOADED_PRGS = {}
local gfx = {}
local gfxSize = ( win.width * win.height )-1
local currKeyPressed = ""
local currKeyReleased = ""
local currMousePressed = 0
local currMouseReleased = 0
local progs = {}
local wins = win.scale
local winss = 1/win.scale
local mousegetX = love.mouse.getX
local mousegetY = love.mouse.getY
local lfgi = love.filesystem.getInfo
local lfgdi = love.filesystem.getDirectoryItems
local osType = "computer"
local function determineSize( dir )
	local size = 0
	local files = lfgdi( dir )
	for i=1, #files do
		local info = lfgi( dir..files[ i ] )
		if info then
			if info.type == "directory" then
				size = size + determineSize( dir..files[ i ].."/" )
			elseif info.type == "file" then
				size = size + round( info.size/MAKE_TRUE_SIZE )
			end
		end
	end
	return size
end
local function reset()
	canvas = love.graphics.newCanvas( win.width, win.height, {
		[ "dpiscale" ] = love.graphics.getDPIScale();
		--[ "format" ] = "hdr";
	} )
	--canvasimageData = canvas:newImageData()
	canvas:setFilter( "nearest", "nearest" )
	love.graphics.setCanvas( canvas )
	for i=0, gfxSize do
		gfx[ i ] = {0,0,0}
	end
	local ok, err = love.filesystem.load( "kernel/boot.lua" )
	USED_RAM = USED_RAM + math.floor(love.filesystem.getInfo( "kernel/boot.lua" ).size/MAKE_TRUE_SIZE)
	if not ok then error( err ) else
		setmetatable( G_ENV, {} )
		for k, v in pairs( _G ) do
			if k ~= "love" and k ~= "G_ENV" then
				G_ENV[ k ] = v
			end
		end
		G_ENV.__NAME = "boot"
		G_ENV.gfx.width = win.width
		G_ENV.gfx.height = win.height
		setfenv( ok, G_ENV )
		ok, err = pcall( ok )
		if not ok then error( err ) end
	end
	love.mouse.setVisible( false )
	if love.filesystem.getInfo( "root" ) == nil then
		love.filesystem.createDirectory( 'root' )
	end
	love.graphics.setCanvas()
end
local function sliceDirs( input )
	local cm = {}
	local word = ""
	local j = 0
	for i=1, #input do
		local c = input:sub( i, i )
		if c == "/" then
			cm[ #cm + 1 ] = word
			word = ""
			j = i+1
		else
			word = word .. c
		end
	end
	cm[ #cm + 1 ] = input:sub( j )
	return cm
end
G_ENV = {
	[ "gfx" ] = {};
	[ "network" ] = {
		[ "isConnected" ] = function()
			local test = socket.tcp()
			test:settimeout(1000)
			local testResult = test:connect("www.google.com", 80)
			test:close()
			test = nil
			if not(testResult == nil) then
				return true
			else
				return false
			end
		end;
	};
	[ "credits" ] = function()
		return CREDITS
	end;
	[ "fs" ] = {
		[ "getName" ] = function( path )
			local lastIndex = 1
			for i=1, #path do
				local c=path:sub( i, i )
				if c == "/" then
					lastIndex = i+1
				end
			end
			return path:sub( lastIndex, #path )
		end;
		[ "mkdir" ] = function( path )
			if path:sub( 1, 1 ) == "/" then
				if #path > 2 then
					path = path:sub( 2 )
				else
					path = ""
				end
			end
			love.filesystem.createDirectory( "root/"..path )
		end;
		[ "exists" ] = function( path )
			local opath = "root/"
			if path:sub( 1, 1 ) == "/" then
				if #path > 2 then
					path = path:sub( 2 )
				else
					path = ""
				end
			end
			local original = sliceDirs( path )[ 1 ]
			if original == "rom" then
				opath = "kernel/"
			end
			if love.filesystem.getInfo( opath..path ) == nil then
				return false
			else
				return true
			end
		end;
		[ "getInfo" ] = function( path )
			local opath = "root/"
			if path:sub( 1, 1 ) == "/" then
				if #path > 2 then
					path = path:sub( 2 )
				else
					path = ""
				end
			end
			local original = sliceDirs( path )[ 1 ]
			if original == "rom" then
				opath = "kernel/"
			end
			local info = love.filesystem.getInfo( opath..path )
			if info then
				if info.type == "directory" then
					info.size = determineSize( opath..path.."/" )
				else
					if info.size then
						info.size = round( info.size / MAKE_TRUE_SIZE )
					end
				end
			end
			return info
		end;
		[ "list" ] = function( path )
			local opath = "root/"
			if path:sub( 1, 1 ) == "/" then
				if #path > 2 then
					path = path:sub( 2 )
				else
					path = ""
				end
			end
			local original = sliceDirs( path )[ 1 ]
			if original == "rom" then
				opath = "kernel/"
			end
			local files = love.filesystem.getDirectoryItems( opath..path )
			if path == "" then
				files[ #files + 1 ] = "rom"
			end
			return files
		end;
		[ "open" ] = function( path, rt )
			local opath = "root/"
			if path:sub( 1, 1 ) == "/" then
				if #path > 2 then
					path = path:sub( 2 )
				else
					path = ""
				end
			end
			if rt == 'r' then
				local original = sliceDirs( path )[ 1 ]
				if original == "rom" then
					opath = "kernel/"
				end
			end
			return love.filesystem.newFile( opath..path, rt )
		end;
		[ "read" ] = function( path )
			local opath = "root/"
			if path:sub( 1, 1 ) == "/" then
				if #path > 2 then
					path = path:sub( 2 )
				else
					path = ""
				end
			end
			local original = sliceDirs( path )[ 1 ]
			if original == "rom" then
				opath = "kernel/"
			end
			return love.filesystem.read( opath..path )
		end;
		[ "write" ] = function( path, value )
			if path:sub( 1, 1 ) == "/" then
				if #path > 2 then
					path = path:sub( 2 )
				else
					path = ""
				end
			end
			return love.filesystem.write( "root/"..path, value )
		end;
		[ "delete" ] = function( path )
			if path:sub( 1, 1 ) == "/" then
				if #path > 2 then
					path = path:sub( 2 )
				else
					path = ""
				end
			end
			love.filesystem.remove( "root/"..path )
		end;
	};
	[ "program" ] = {
		[ "load" ] = function( path, loadedBy )
			local opath = "root/"
			if path:sub( 1, 1 ) == "/" then
				if #path > 2 then
					path = path:sub( 2 )
				else
					path = ""
				end
			end
			local original = sliceDirs( path )[ 1 ]
			if original == "rom" then
				opath = "kernel/"
			end
			local tbl, err = love.filesystem.getInfo( opath..path )
			local size = 0
			if tbl then
				if tbl.size >= MAKE_TRUE_SIZE then
					size = math.floor(tbl.size/MAKE_TRUE_SIZE)
				end
			else
				return false, "No such program '"..path.."'"
			end
			USED_RAM = USED_RAM + size
			if USED_RAM > MAX_RAM then
				USED_RAM = USED_RAM - size
				return false, "Max RAM Exceeded!"
			else
				LOADED_PRGS[ path ] = {
					[ "size" ] = size;
					[ "func" ] = love.filesystem.load( opath..path );
					[ "loadedBy" ] = loadedBy;
					[ "sandbox" ] = {};
				}
				return true
			end
		end;
		[ "unload" ] = function( path )
			if LOADED_PRGS[ path ] then
				USED_RAM = USED_RAM - LOADED_PRGS[ path ].size
				for k, v in pairs( LOADED_PRGS ) do
					if v.loadedBy == path then
						USED_RAM = USED_RAM - v.size
						LOADED_PRGS[ k ] = nil
					end
				end
				LOADED_PRGS[ path ] = nil
				return true
			else
				return false, "Program not loaded!"
			end
		end;
		[ "setfenv" ] = function( path, env )
			if LOADED_PRGS[ path ] then
				LOADED_PRGS[ path ].sandbox = env
				return true
			else
				return false, "Program not loaded!"
			end
		end;
		[ "start" ] = function( path, ... )
			if LOADED_PRGS[ path ] then
				setmetatable( LOADED_PRGS[ path ].sandbox, {} )
				for k, v in pairs( G_ENV ) do
					LOADED_PRGS[ path ].sandbox[ k ] = v
				end
				setfenv( LOADED_PRGS[ path ].func, LOADED_PRGS[ path ].sandbox )
				LOADED_PRGS[ path ].sandbox.__NAME = path
				local ok, err = pcall( LOADED_PRGS[ path ].func, ... )
				return ok, err
			else
				return false, "Program not loaded!"
			end
		end;
		[ "getSize" ] = function( path )
			if LOADED_PRGS[ path ] then
				return LOADED_PRGS[ path ].size
			end
		end;
		[ "getLoadedBy" ] = function( path )
			if LOADED_PRGS[ path ] then
				return LOADED_PRGS[ path ].loadedBy
			end
		end;
		[ "update" ] = function( path, dt )
			if LOADED_PRGS[ path ] then
				local env = getfenv( LOADED_PRGS[ path ].func )
				if env.update then
					env.update( dt )
				end
			end
		end;
	};
	[ "sys" ] = {
		[ "filterBlue" ] = function( amount )
			if amount > 0 and amount <= 1 then
				blueCutoff = floor( blueAmount*amount )
			else
				blueCutoff = 0
			end
		end;
		[ "setScreenMode" ] = function( mode )
			if mode == "low" then
				if osType == "pocket" then
					win.scale = 1
					wins = 1
					winss = 1
				else
					win.scale = 3
					wins = 3
					winss = 1/3
				end
			elseif mode == "high" then
				if osType == "pocket" then
					win.scale = 3
					wins = 3
					winss = 1/3
				else
					win.scale = 5
					wins = 5
					winss = 1/5
				end
			else
				if osType == "pocket" then
					win.scale = 2
					wins = 2
					winss = 1/2
				else
					win.scale = 4
					wins = 4
					winss = 1/4
				end
			end
			love.window.setMode(0, 0)
			screen_xoff = love.graphics.getWidth()/2-(win.width*win.scale/2)
			screen_yoff = love.graphics.getHeight()/2-(win.height*win.scale/2)
			success = love.window.setMode( round(love.graphics.getWidth()/win.scale), round(love.graphics.getHeight()/win.scale), {
				[ "vsync" ] = false;
				[ "stencil" ] = false;
				[ "borderless" ] = true;
				[ "fullscreen" ] = true;
			} )
			love.graphics.setPointSize( win.scale )
		end;
		[ "getRamSize" ] = function()
			return MAX_RAM
		end;
		[ "getRamUsed" ] = function()
			return USED_RAM
		end;
		[ "getStorageSize" ] = function()
			return MAX_STORAGE
		end;
		[ "getUsedStorage" ] = function()
			return USED_STORAGE
		end;
		[ "getCPUSpeed" ] = function()
			return CPU_SPEED
		end;
		[ "textInput" ] = "";
		[ "keyDown" ] = love.keyboard.isDown;
		[ "getMousePos" ] = function()
			if osType == "pocket" then
				return floor((mousegetX()-screen_xoff) * winss), floor((mousegetY()-screen_yoff) * winss)
			else
				return floor(mousegetX() * winss), floor(mousegetY() * winss)
			end
		end;
		[ "mouseIsDown" ] = love.mouse.isDown;
		[ "mousePressed" ] = function( key )
			if currMousePressed == key then
				currMousePressed = 0
				return true
			end
			return false
		end;
		[ "reset" ] = reset;
		[ "mouseReleased" ] = function( key )
			if currMouseReleased == key then
				currMouseReleased = 0
				return true
			end
			return false
		end;
		[ "keyPressed" ] = function( key )
			if currKeyPressed == key then
				return true
			end
			return false
		end;
		[ "keyReleased" ] = function( key )
			if currKeyReleased == key then
				return true
			end
			return false
		end;
		[ "setKeyRepeat" ] = love.keyboard.setKeyRepeat;
		[ "getKeyRepeat" ] = love.keyboard.hasKeyRepeat;
		[ "getMachineType" ] = function()
			return osType
		end;
	}
}

function love.textinput( t )
	G_ENV.sys.textInput = G_ENV.sys.textInput .. t
end

function love.keypressed(key)
	currKeyPressed = key
end

function love.keyreleased(key)
	currKeyReleased = key
end

function love.wheelmoved(x, y)
	if y > 0 then
		currMousePressed = "wu"
		currMouseReleased = "wd"
	elseif y < 0 then
		currMousePressed = "wd"
		currMouseReleased = "wu"
    end
end

function love.mousepressed(_,_,key)
	currMousePressed = key
end

function love.mousereleased(_,_,key)
	currMouseReleased = key
end

local setColor = love.graphics.setColor
local point = love.graphics.points
local clear = love.graphics.clear
local ccr = 0
local ccg = 0
local ccb = 0
local csr = 0
local csg = 0
local csb = 0
G_ENV.gfx.clear = function( rgb )
	local rgb = rgb or { 0x0, 0x0, 0x0 }
	for i=0, gfxSize do
		gfx[ i ] = rgb
	end
	if csr ~= rgb[1] then
		csr = rgb[1]
		ccr = (rgb[1]*17)*twofivefive
	end
	if csg ~= rgb[2] then
		csg = rgb[2]
		ccg = (rgb[2]*17)*twofivefive
	end
	if csb ~= rgb[3] then
		csb = rgb[3]
		ccb = (rgb[3]*17)*twofivefive
	end
	clear( ccr, ccg, ccb )
end
math.round=round
local winw = win.width
local winww = 1/win.width
local winsh = wins/2
local pcr = 0
local pcg = 0
local pcb = 0
local ppr = 0
local ppg = 0
local ppb = 0
G_ENV.gfx.putPixel = function( x, y, rgb )
	x=round(x)
	y=round(y)
	if x >= 0 and x < winw and y>=0 and y<win.height then
		rgb = rgb or { 0xF, 0xF, 0xF }
		if ppr ~= rgb[1] then
			ppr = rgb[1]
			pcr = (rgb[1]*17)*twofivefive
		end
		if ppg ~= rgb[2] then
			ppg = rgb[2]
			pcg = (rgb[2]*17)*twofivefive
		end
		if ppb ~= rgb[3] then
			ppb = rgb[3]
			pcb = (rgb[3]*17)*twofivefive
		end
		gfx[ y*winw+x ] = rgb
		setColor( pcr, pcg, pcb )
		point( 1+x, 1+y )
	end
end
local one17=1/17
G_ENV.gfx.getPixel = function( x, y )
	x=round(x)
	y=round(y)
	if x >= 0 and x < winw and y>=0 and y<win.height then
		return gfx[ y*winw+x ]
	end
end

function love.load()
	if love.system.getOS() == "Android" then
		osType = "pocket"
		love.keyboard.setTextInput( true )
		love.keyboard.setKeyRepeat( true )
		win.scale = 2
		wins = 2
		winss = 1/2
	end
	love.filesystem.setIdentity( "Neon Machine", false )
	love.window.setMode(0, 0)
	screen_xoff = love.graphics.getWidth()/2-(win.width*win.scale/2)
	screen_yoff = love.graphics.getHeight()/2-(win.height*win.scale/2)
	if osType ~= "pocket" then
		success = love.window.setMode( round(love.graphics.getWidth()/win.scale), round(love.graphics.getHeight()/win.scale), {
			[ "vsync" ] = false;
			[ "stencil" ] = false;
			[ "borderless" ] = true;
			[ "fullscreen" ] = true;
		} )
	else success = true
	end
	--love.graphics.setPointSize( win.scale )
	if( success ) then
		reset()
	end
	love.graphics.setCanvas()
end

local mydt = 0
local counter = 0
local genvu = G_ENV.update
local setCanvas = love.graphics.setCanvas
function love.update( dt )
	if osType == "pocket" then
		local touches = love.touch.getTouches()
		for i, id in ipairs(touches) do
			local x, y=love.touch.getPosition(id)
			if x < screen_xoff or x > screen_xoff + (canvas:getWidth()*win.scale) then
				if love.keyboard.hasTextInput() == false then
					love.keyboard.setTextInput( true )
				end
			else
				if love.keyboard.hasTextInput() == true then
					love.keyboard.setTextInput( false )
				end
			end
		end
	end
	if genvu then
		setCanvas( canvas )
		local ok, err = pcall( genvu, dt )
		if not ok then error( err ) end
		setCanvas()
		mydt = mydt + dt
		if currKeyReleased ~= '' then
		 currKeyReleased = ""
		end
		if currKeyPressed ~= '' then
		 currKeyPressed = ""
		end
		currMouseReleased = 0
		currMousePressed = 0
	else
		genvu = G_ENV.update
	end
	if counter >= 60 then
		CPU_SPEED = mydt
		mydt = 0
		counter = 0
		USED_STORAGE = determineSize( 'root/' )
		if winw ~= win.width then
			winw = win.width
			winww = 1/win.width
		end
	end
	counter = counter + 1
end

local floor = math.floor
local draw = love.graphics.draw
function love.draw()
	setColor( 1, 1, 1 )
	draw( canvas, screen_xoff, screen_yoff, 0, wins, wins )
end








